core/src/main/java/org/apache/accumulo/core/data/Value.java

```diff
@@ -181,6 +182,15 @@ public void readFields(final DataInput in) throws IOException {
     in.readFully(this.value, 0, this.value.length);
   }
 
+  public void readFields(final DataInput in, Supplier<Long> freeMemorySupplier) throws IOException {
+    int length = in.readInt();
+    while (freeMemorySupplier.get() < length) {
+      Thread.onSpinWait();
+    }
+    this.value = new byte[length];
```

Comment:
I haven't fully thought about this approach yet in terms of the consequences of trying to wait for free memory as I've never tried to do it before but one thing that stands out to me is that there seems like there's always going to be a race condition here. Even if you wait and get the ok that there is memory to proceed by the time you continue and get to this line and go to allocate memory for the byte array there's no guarantee the JVM hasn't already allocated some of the memory you'd need for this to work in another thread. In fact, if you were right on the edge and barely had enough memory then I would think this could fail often with the JVM allocating memory in other threads. In practice though I don't know if that's true without testing but it seems a bit non deterministic since you don't have much control over the GC process and what other threads are doing.

Reply:
I agree that it's not a perfect solution, it's an attempt at doing something. There are other options that could be added, like only check for free memory if the length is over some size. The problem we run into is that the tablet server is running N scans concurrently, JVM memory is under pressure, and it doesn't know that the next K/V it's going to read is going to create a very large (500MB, 1GB) Value object.

Reply:
Maybe we need to have some configurable size to give us a bit of a buffer that we add to the check that could help reduce risk of failures due to race conditions. So instead of just continuing if `free_memory` is large enough, we don't continue until `free_memory + some_configurable_size` (could be a meg, 10 megs, whatever) of memory is available to reduce the risk of failures.

Reply:
> There are other options that could be added, like only check for free memory if the length is over some size.
> So instead of just continuing if free_memory is large enough, we don't continue until free_memory + some_configurable_size

I added both of these things in 0177dbd. The default size threshold to perform the free memory check is 5% of max memory. The buffer size is 2% of the Value size.